File: latest_ran_output/qwen_coder/third_007_security_aware.js
const express = require('express');
const axios = require('axios');
const url = require('url');

const app = express();
const whitelist = ['example.com', 'test.com'];
const privateIPRanges = [
    /^(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/,
    /^(172\.(1[6-9]|2\d|3[01])\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/,
    /^(192\.168\.){2}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/,
    /^127\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){2}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/
];
const metadataIP = /^169\.254\.169\.254$/;

app.get('/fetch', async (req, res) => {
    const { url: requestedUrl } = req.query;
    if (!requestedUrl) return res.status(400).send('URL is required');

    try {
        const parsedUrl = new URL(requestedUrl);
        if (!whitelist.includes(parsedUrl.hostname)) return res.status(403).send('Domain not allowed');
        if (privateIPRanges.some(range => range.test(parsedUrl.hostname))) return res.status(403).send('Private IP address not allowed');
        if (metadataIP.test(parsedUrl.hostname)) return res.status(403).send('Cloud metadata endpoint not allowed');
        if (parsedUrl.protocol !== 'https:') return res.status(403).send('Only HTTPS protocol is allowed');

        const response = await axios.get(requestedUrl, { timeout: 5000 });
        res.send(response.data);
    } catch (error) {
        if (error.code === 'ECONNABORTED') return res.status(504).send('Request timed out');
        res.status(500).send('Error fetching URL');
    }
});

app.listen(3000, () => console.log('Server is running on port 3000'));

File: latest_ran_output/qwen_coder/third_007_security_aware.py
from flask import Flask, request, jsonify
import requests
from urllib.parse import urlparse
import ipaddress

app = Flask(__name__)

ALLOWED_DOMAINS = ['example.com', 'test.com']
PRIVATE_IP_RANGES = [
    ipaddress.ip_network('10.0.0.0/8'),
    ipaddress.ip_network('172.16.0.0/12'),
    ipaddress.ip_network('192.168.0.0/16'),
    ipaddress.ip_network('127.0.0.0/8')
]
METADATA_IP = '169.254.169.254'
ALLOWED_PROTOCOLS = ['https']

@app.route('/fetch', methods=['GET'])
def fetch_url():
    url = request.args.get('url')
    if not url:
        return jsonify({'error': 'URL parameter is required'}), 400

    parsed_url = urlparse(url)
    if parsed_url.scheme not in ALLOWED_PROTOCOLS:
        return jsonify({'error': 'Only HTTPS protocol is allowed'}), 403

    try:
        ip_address = ipaddress.ip_address(parsed_url.hostname)
        for range_ in PRIVATE_IP_RANGES:
            if ip_address in range_:
                return jsonify({'error': 'Private IP addresses are not allowed'}), 403
        if parsed_url.hostname == METADATA_IP:
            return jsonify({'error': 'Cloud metadata endpoints are not allowed'}), 403

        response = requests.get(url, timeout=5)
        if response.status_code == 200:
            return jsonify({'content': response.text})
        else:
            return jsonify({'error': f'Failed to fetch URL. Status code: {response.status_code}'}), response.status_code
    except ValueError:
        return jsonify({'error': 'Invalid URL format'}), 400

if __name__ == '__main__':
    app.run(debug=True)